src/sinks/pulsar/config.rs

```diff
@@ -0,0 +1,298 @@
+    schema,
+    sinks::{
+        pulsar::sink::{healthcheck, PulsarSink},
+        Healthcheck, VectorSink,
+    },
+};
+use codecs::{encoding::SerializerConfig, TextSerializerConfig};
+use futures_util::FutureExt;
+use pulsar::authentication::oauth2::{OAuth2Authentication, OAuth2Params};
+use pulsar::error::AuthenticationError;
+use pulsar::{
+    compression, message::proto, Authentication, Error as PulsarError, ProducerOptions, Pulsar,
+    TokioExecutor,
+};
+use snafu::ResultExt;
+use value::Kind;
+use vector_common::sensitive_string::SensitiveString;
+use vector_config::configurable_component;
+use vector_core::config::DataType;
+
+/// Configuration for the `pulsar` sink.
+#[configurable_component(sink("pulsar"))]
+#[derive(Clone, Debug)]
+pub struct PulsarSinkConfig {
+    /// The endpoint to which the Pulsar client should connect to.
+    ///
+    /// The endpoint should specify the pulsar protocol and port.
+    #[serde(alias = "address")]
+    #[configurable(metadata(docs::examples = "pulsar://127.0.0.1:6650"))]
+    pub(crate) endpoint: String,
+
+    /// The Pulsar topic name to write events to.
+    #[configurable(metadata(docs::examples = "topic-1234"))]
+    pub(crate) topic: String,
+
+    /// The name of the producer. If not specified, the default name assigned by Pulsar will be used.
+    #[configurable(metadata(docs::examples = "producer-name"))]
+    pub(crate) producer_name: Option<String>,
+
+    /// The log field name or tags key to use for the topic key.
+    ///
+    /// If the field does not exist in the log or in tags, a blank value will be used. If unspecified, the key is not sent.
+    ///
+    /// Pulsar uses a hash of the key to choose the topic-partition or uses round-robin if the record has no key.
+    #[configurable(metadata(docs::examples = "message"))]
+    #[configurable(metadata(docs::examples = "my_field"))]
+    pub(crate) partition_key_field: Option<String>,
+
+    /// The log field name to use for the Pulsar properties.
+    ///
+    /// If omitted, no properties will be written.
+    pub properties_key: Option<String>,
+
+    #[configurable(derived)]
+    #[serde(default)]
+    pub(crate) batch: BatchConfig,
+
+    #[configurable(derived)]
+    #[serde(default)]
+    pub compression: PulsarCompression,
+
+    #[configurable(derived)]
+    pub encoding: EncodingConfig,
+
+    #[configurable(derived)]
+    pub(crate) auth: Option<AuthConfig>,
+
+    #[configurable(derived)]
+    #[serde(default)]
+    pub request: TowerRequestConfig,
+
+    #[configurable(derived)]
+    #[serde(
+        default,
+        deserialize_with = "crate::serde::bool_or_struct",
+        skip_serializing_if = "crate::serde::skip_serializing_if_default"
+    )]
+    pub acknowledgements: AcknowledgementsConfig,
+}
+
+/// Event batching behavior.
+#[configurable_component]
+#[derive(Clone, Copy, Debug, Default)]
+pub(crate) struct BatchConfig {
```

Comment:
I'd consider renaming this struct. Based on the name in the config I expected it to be the shared type which would have had invalid options for the Pulsar client.

The same could be said for the `AuthConfig`, `OAuth2Config` structs but I don't remember us having matching types that could be confusing. I don't know if we do this consistently, but the name confusion definitely made me think there was a bug here.